docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+
+```csharp
+// Ensure that all variables in the local declaration have initializers that
+// are assigned with constant values.
+foreach (var variable in localDeclaration.Declaration.Variables)
+{
+    var initializer = variable.Initializer;
+    if (initializer == null)
+    {
+        return;
+    }
+
+    var constantValue = context.SemanticModel.GetConstantValue(initializer.Value);
+    if (!constantValue.HasValue)
+    {
+        return;
+    }
+}
+```
+
+The first `if` in the preceding code exits as soon one variable declaration that doesn't have an initializer was found. The second `if` returns when a declaration has an initializer that is not a constant.
+
+At this point, your code has found a declaration that is initialized with a constant. The declaration can be made constant provided its value is not modified after it has been initialized. You can illustrate that by creating one more test:
+
+[!code-csharp[Don't report diagnostics when variable is assigned to](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#VariableChangesValue "Test to ensure no diagnostics are reported when the variable declaration is assigned to")]
+
+### Make sure the variable isn't assigned after initialization
+
+You'll use the <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> to perform data flow analysis on the local declaration statement. Then, you use the results of this data flow analysis to ensure that none of the local variables are written with a new value anywhere else. Call the <xref:Microsoft.CodeAnalysis.ModelExtensions.GetDeclaredSymbol%2A> extension method to retrieve the <xref:Microsoft.CodeAnalysis.ILocalSymbol> for each variable and checking that it isn't contained with the <xref:Microsoft.CodeAnalysis.DataFlowAnalysis.WrittenOutside%2A?displayProperty=nameWithType> collection of the data flow analysis. Add the following code to `AnalyzeNode` after the closing brace of the `foreach` loop:
+
+[!code-csharp[Use flow analysis to find writes](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst/MakeConstAnalyzer.cs#FlowAnalysis "Use flow analysis to detect writes")]
+
+### Create and report the diagnostic
+
+You're almost done. You've written the code to find a local declaration, and filtered out cases where a local declaration cannot be const. Now, let's write a test the shows a declaration that should be const. Open **MakeConstUnitTests.cs** and change `TestMethod2` to the following test:
```

Comment:
Would it be easier to have them delete `TestMethod1` and `TestMethod2` at the start?